Homework 3

Falltren opened the pull request.

Falltren committed:
добавлены сервлеты для обработки запросов, связанных с пользователями

Falltren committed:
добавлен сервлет обрабатывающий запрос на создание привычки
добавлен маппер для конвертации сущность/дто (привычки)

Falltren committed:
добавлены сервлеты для выполнения crud операций с привычками, а также подтверждения выполнения привычки

Falltren committed:
добавлено использование аспектов для замера времени выполнения методов

Falltren committed:
добавлены тесты сервлетов

Falltren committed:
обновлены тесты сервисов

Falltren committed:
добавлены тесты для сервиса валидации дто, исправлены баги в работе сервиса валидации

Falltren committed:
добавлен сервлет для получения отчета, а также тесты, покрывающие валидацию дто/сервис/сервлет

Falltren committed:
добавлен аудит действий пользователя

Falltren committed:
обновлены javadoc и readme

Falltren committed:
удаление неиспользуемых классов

YlabCourse commented on tracking_habits/src/main/java/com/fallt/aop/logging/LoggableAspect.java:
```diff
@@ -0,0 +1,25 @@
+        long start = System.currentTimeMillis();
+        Object result = proceedingJoinPoint.proceed();
+        long time = System.currentTimeMillis() - start;
+        System.out.println("Выполнение метода " + methodName + " завершено. Время выполнения метода составило: " + time + " мс.");
```
Логирование должно быть реализовано за счет Slf4j, а не вывода в консоль.

YlabCourse commented on tracking_habits/src/main/java/com/fallt/dto/request/ReportRequest.java:
```diff
@@ -0,0 +1,21 @@
+
+    private LocalDate start;
+
+    private LocalDate end;
```
Есть смысл ставить пустые строки между методами, после начала класса, между полями которые имеют аннотацию или java doc, или семантически принадлежат к разной логике, здесь в этом нет особого смысла.

YlabCourse commented on tracking_habits/src/main/java/com/fallt/aop/audit/Audit.java:
Этот файл должен быть в пакете dto с неймингом AuditDto или AuditRequest.

YlabCourse commented on tracking_habits/src/main/java/com/fallt/aop/audit/AuditAspect.java:
```diff
@@ -0,0 +1,111 @@
+        String userEmail = parameter.getEmail();
+        Audit audit = new Audit("Используемый email " + userEmail, "Регистрация пользователя", LocalDateTime.now());
+        writer.write(audit);
+    }
```
Ты бы мог просто сделать общий аспект для всех методов помеченных `@Auditable` где ты бы вытаскивал через
`ProceedingJoinPoint` имя метода, а из аннотации нужную тебе информацию.

YlabCourse commented on tracking_habits/src/main/java/com/fallt/aop/audit/AuditAspect.java:
```diff
@@ -0,0 +1,111 @@
+    public void beforeUpdateMethod(JoinPoint joinPoint) {
+        String userEmail = (String) joinPoint.getArgs()[0];
+        Audit audit = new Audit("Используемы email " + userEmail, "Обновление данных о пользователе", LocalDateTime.now());
+        writer.write(audit);
```
При аудировании следует сохранять аудит в базу данных, а не просто выводить в консоли.

YlabCourse commented on tracking_habits/src/main/java/com/fallt/dto/response/ExecutionDto.java:
Зачем здесь переопределять toString если у тебя аннотация Data

YlabCourse commented on tracking_habits/src/main/java/com/fallt/mapper/HabitMapper.java:
```diff
@@ -0,0 +1,31 @@
+
+    HabitMapper INSTANCE = getMapper(HabitMapper.class);
+
+    @Mapping(target = "createAt", expression = "java(java.time.LocalDate.now())")
```
Это можно убрать в сущность за счет аннотацию `@Builder.Default` и значением которое будет инициализировано по дефолту, в твоем случае это LocalDate.now.

YlabCourse commented on tracking_habits/src/main/java/com/fallt/mapper/UserMapper.java:
```diff
@@ -0,0 +1,26 @@
+import com.fallt.dto.request.UpsertUserRequest;
+import com.fallt.dto.response.UserResponse;
+import com.fallt.entity.User;
+import org.mapstruct.*;
```
В предыдущем дз было замечание насчет wildcard.

YlabCourse commented on tracking_habits/src/main/java/com/fallt/security/AuthenticationContext.java:
```diff
@@ -0,0 +1,58 @@
+        checkAuthentication(userEmail);
+        Role currentUserRole = context.get(userEmail).getRole();
+        if (!currentUserRole.equals(requiredRole)) {
+            throw new SecurityException("У вас недостаточно прав для выполнения данного действия"); // при использовании spring будет приводить к статусу 403 в ответе
```
1. Не должно быть таких комментов в мре.
2. А что сейчас мешает 403 в ответе, ты же можешь возвращать в контроллере 403 если currentUserRole!=requiredRole.

YlabCourse commented on tracking_habits/src/main/java/com/fallt/service/AuthService.java:
```diff
@@ -1,37 +1,44 @@
 @RequiredArgsConstructor
+@Loggable
+@Auditable
 public class AuthService {
```
Должен быть интерфейс который будет представлять абстракцию над классом имплементации(Этим классом), в котором и будет написан javadoc.

YlabCourse commented on tracking_habits/src/main/java/com/fallt/service/AuthService.java:
```diff
@@ -1,37 +1,44 @@
-    private final ConsoleOutput consoleOutput;
-
     /**
      * Проверка наличия пользователя в базе данных
```
Такие комментарии не являются полезными, тк просто описывает, то что и так  понятно по названию метода.

YlabCourse commented on tracking_habits/src/main/java/com/fallt/service/HabitService.java:
```diff
@@ -1,110 +1,118 @@
         if (optionalHabit.isEmpty()) {
-            consoleOutput.printMessage(Message.INCORRECT_HABIT_TITLE);
-            return;
+            throw new EntityNotFoundException(MessageFormat.format("У вас отсутствует привычка с указанным названием: {0}", title));
```
А почему юзер не может обновить идею, оставив то же название?

YlabCourse commented on tracking_habits/src/main/java/com/fallt/service/HabitService.java:
```diff
@@ -1,110 +1,118 @@
-    public void updateHabit(User user, String title, HabitDto dto) {
+    public HabitResponse updateHabit(String userEmail, String title, UpsertHabitRequest request) {
+        User user = userService.getUserByEmail(userEmail);
         Optional<Habit> optionalHabit = findHabit(user, title);
```
Почему привычка для обновления ищется по двум параметрам юзер и тайтл, при том что в методе от юзера нужен только юзерайди.

Да и в целом обновление должно выглядеть так, что ты не вынимаешь снова из бд  нужную тебе привычку и обновляешь ее, а вводишь айдишник привычки и дто которая содержит новые данные для обновления.

YlabCourse commented on tracking_habits/src/main/java/com/fallt/service/HabitService.java:
```diff
@@ -1,110 +1,118 @@
+        return response;
     }
 
     private Optional<Habit> findHabit(User user, String title) {
```
Стоит в этом методе пробрасывать EntityNotFoundException через orElseThrow, плюс такие методы всегда находятся в самом ниху обособленно от основной логики.

YlabCourse commented on tracking_habits/src/main/java/com/fallt/service/ValidationService.java:
Валидация не должна проводится на уровне контроллеров, это должно происходить в сервисном слое, а у тебя этот класс используется в контроллерах.

YlabCourse commented on tracking_habits/src/main/java/com/fallt/servlet/AdminServlet.java:
```diff
@@ -0,0 +1,69 @@
+@WebServlet("/admin")
+public class AdminServlet extends HttpServlet {
+
+    private static final String OBJECT_MAPPER = "objectMapper";
+    private static final String USER_SERVICE = "userService";
+    private static final String AUTH_CONTEXT = "authContext";
```
Это стоит отделить в класс с константами, а не хранить в контроллере.

YlabCourse commented on tracking_habits/src/main/java/com/fallt/servlet/AdminServlet.java:
```diff
@@ -0,0 +1,69 @@
+    @Override
+    protected void doGet(HttpServletRequest req, HttpServletResponse resp) throws ServletException, IOException {
+        ServletContext context = getServletContext();
+        ObjectMapper objectMapper = (ObjectMapper) context.getAttribute(OBJECT_MAPPER);
+        UserService userService = (UserService) context.getAttribute(USER_SERVICE);
```
Ты можешь делать это в методе init унаследованный от HttpServlet.

YlabCourse commented on tracking_habits/src/test/java/com/fallt/service/UserServiceTest.java:
```diff
@@ -1,61 +1,64 @@
-    void testGetAllUsers(){
-        List<User> habits = List.of(
+    void testGetAllUsers() {
+        List<User> users = List.of(
                 User.builder().name("user1").password("pwd1").email("email1").build(),
                 User.builder().name("user2").password("pwd2").email("email2").build()
         );
-        when(userDao.findAll()).thenReturn(habits);
+        List<UserResponse> expected = List.of(
+                new UserResponse("user1", "email1"),
+                new UserResponse("user2", "email2")
+        );
+        when(userDao.findAll()).thenReturn(users);
```
Тестовые значения стоит вынести в константы.

YlabCourse commented on tracking_habits/src/test/java/com/fallt/service/UserServiceTest.java:
```diff
@@ -79,123 +82,113 @@ void testGetUserByIncorrectEmail() {
     }
 
-    private User getUserFromDatabase() {
+    private User createUser() {
```
Эти методы не нужны, если заменить значения на константы.

YlabCourse reviewed (commented):
Есть недочеты:

- Много ошибок в архитектуре, часто вижу что методы используются и находятся не там где нужно.
- Есть логические ошибки в написании методов.
- Комментарии хотелось бы видеть более содержательными.
- Нет классов с константами.

Falltren committed:
добавлено логирование

Falltren committed:
откорректирована работа аудита -> вывод в консоль заменен на сохранение в бд

Falltren committed:
Добавлен класс с константами, удален хардкодинг значений
Получение инстантов классов вынесено из методов сервлетов, отвечающих за обработку запросов в init метод

Falltren committed:
Добавлены интерфейсы для сервисного слоя
